PHP 8 compatibility

bennothommo opened the pull request:
This is an open PR towards PHP 8 compatibility in October CMS.

It is not recommended to use October CMS in a PHP 8 environment at this time. Please stick to PHP 7.3 and 7.4, which are the currently supported PHP versions of October CMS that are still receiving updates from the PHP development team.

Rain Library PR: https://github.com/octobercms/library/pull/558

Big thanks to @tolkamsoft for their assistance with this so far.

Commit:
Run tests on PHP 8

These tests will be allowed to fail for now, as we don't officially support PHP 8.

Commit:
Fix YAML formatting

Commit:
Remove second include

Commit:
Further fixes

ayumi-cloud commented:
Interesting! 

I seen you wrote `phpVersion: '8.0'` 

I've seen other repo's write this: `phpVersion: 'nightly'` instead, for php 8.x

@bennothommo When you free, could you tell me why all the pull requests are failing here? (I see the last few pr's are all failing with the same issue).

marcsoler commented:
PHP 8 is oficially out and as far as I understand, doctrine/dbal latest version (3.0) now supports PHP 8. [october/rain](octobercms/library) currently requires dbal 2.6.

bennothommo commented:
@ayumi-cloud `nightly` would be PHP 8.1 at this point, and given that GitHub Actions doesn't have an "allow failure" type of action, we'd probably constantly get failures if we tracked `nightly`. I'd rather us just focus our efforts on the supported versions, as opposed to chasing a white unicorn like `nightly` releases.

ayumi-cloud commented:
Yah good point about `8.1` didn't think of that.

`white unicorn releases` haha made me laugh.

bennothommo commented:
![image](https://i.pinimg.com/originals/75/7e/6d/757e6d0ad29432d76cff269c2f514c64.gif)

iotch commented:
Seems like PHP 8.0 tests are failing on `Doctrine\DBAL\Driver\PDOConnection::query()` however Doctrine DBAL is compatible with php 8 and just needs to be updated to v2.12.

Commit:
Merge branch 'develop' into wip/php-8-testing

Commit:
PHP 8 compatibility fixes (#5454)

Co-authored-by: Kanstantsin K <kk@tolkam.by>

Commit:
Fixed call_user_func_array() named parameters mismatch that came from BackendController (#5493)

Co-authored-by: Kanstantsin K <kk@tolkam.by>

Commit:
Merge remote-tracking branch 'origin/develop' into wip/php-8-testing

Commit:
Force PHP version 7.2.9 as per PHP 8 changes to library

Commit:
Remove platform lock for PHP version.

Since we are now relying on different versions of some dependencies depending on whether PHP 7 or 8 is used, we can no longer lock the PHP version down.

Commit:
Adjust dev dependencies

- No longer need AssertGD in main repo, as it was only used for library tests.
- Allow different version of ArraySubset asserts depending on PHP version

Commit:
Add stubs for renamed PHPUnit methods in PHPUnit 9

Commit:
Test PHP 8 on both Linux and Windows

Commit:
Remove experimental flag

bennothommo commented:
@LukeTowers this is ready for final review. Unit tests appear to be working with PHP 8, but I haven't yet tested October directly on PHP 8.

LukeTowers commented on composer.json:
```diff
@@ -51,8 +51,7 @@
         "fzaninotto/faker": "~1.9",
         "squizlabs/php_codesniffer": "3.*",
         "php-parallel-lint/php-parallel-lint": "^1.0",
-        "meyfa/phpunit-assert-gd": "^2.0.0",
```
I'm assuming this is removed because it's included in the library test suite?

bennothommo replied:
Correct :)

LukeTowers commented on composer.json:
```diff
@@ -82,10 +81,7 @@
     "config": {
-        "preferred-install": "dist",
-        "platform": {
-            "php": "7.2.5"
```
This is to prevent differences between the marketplace delivered version and the composer one, but perhaps it's finally been fixed on the marketplace side by @daftspunk?

LukeTowers commented on modules/backend/classes/BackendController.php:
```diff
@@ -111,6 +111,14 @@ public static function extend(Closure $callback)
+    public function callAction($method, $parameters)
+    {
+        return parent::callAction($method, array_values($parameters));
+    }
```
Why was this required? Has Laravel not fixed it on their end?

tolkamsoft replied:
I believe this is due to Laravel is using array destructing and in php 8 array keys should match parameter names during destructing, so 0, 1, etc. are not valid parameters names in this case.

LukeTowers commented on modules/backend/classes/BackendController.php:
```diff
@@ -210,7 +218,7 @@ protected function getRequestedController($url)
          */
         if (count($params) >= 2) {
-            list($author, $plugin) = $params;
+            [$author, $plugin] = $params;
```
Has list() been removed or is this just a coding style preference?

iotch replied:
This is just coding style applied by IDE, sorry

tolkamsoft replied:
This is just coding style applied by IDE, sorry
